packages/did/src/didAuthorizeExtrinsics/didAuthorizeExtrinsics.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * Copyright (c) 2018-2022, BOTLabs GmbH.
+ *
+ * This source code is licensed under the BSD 4-Clause "Original" license
+ * found in the LICENSE file in the root directory of this source tree.
+ */
+
+import type { Extrinsic } from '@polkadot/types/interfaces'
+import type { SubmittableExtrinsicFunction } from '@polkadot/api/types'
+
+import { BN } from '@polkadot/util'
+import type {
+  DidVerificationKey,
+  IIdentity,
+  SignCallback,
+  SubmittableExtrinsic,
+  VerificationKeyRelationship,
+} from '@kiltprotocol/types'
+import { SDKErrors } from '@kiltprotocol/utils'
+
+import { FullDidDetails } from '../DidDetails/FullDidDetails.js'
+import {
+  getKeyRelationshipForExtrinsic,
+  increaseNonce,
+} from '../DidDetails/FullDidDetails.utils.js'
+
+import { generateDidAuthenticatedTx } from '../Did.chain.js'
+import { getSigningAlgorithmForVerificationKeyType } from '../Did.utils.js'
+
+/**
+ * Type of a callback used to select one of the key candidates for a DID to sign a given batch of extrinsics.
+ */
+type SelectKeyCallback = (
+  keys: DidVerificationKey[],
+  batch?: Extrinsic[]
+) => Promise<DidVerificationKey>
+
+async function defaultSelectKey([
+  first,
+]: DidVerificationKey[]): ReturnType<SelectKeyCallback> {
+  return first
+}
+
+type GroupedExtrinsics = Array<{
+  extrinsics: Extrinsic[]
+  keyRelationship: VerificationKeyRelationship
+}>
+
+function groupExtrinsicsByKeyRelationship(
+  extrinsics: Extrinsic[]
+): GroupedExtrinsics {
+  const [first, ...rest] = extrinsics.map((extrinsic) => {
+    const keyRelationship = getKeyRelationshipForExtrinsic(extrinsic)
+    if (keyRelationship === 'paymentAccount') {
+      throw new SDKErrors.DidBuilderError(
+        'Can only batch extrinsics that require a DID signature'
+      )
+    }
+    return { extrinsic, keyRelationship }
+  })
+
+  const groups: GroupedExtrinsics = [
```

Comment:
This might throw with trying to access undefined, when `extrinsics` is an empty array. I think it would be good to check for that.

Reply:
This is not a public function however, and it’s only called with array of 2 or more. Should I still add the check?

Reply:
ah ok, didn't see that. hmm, personally still feels better to have it, just in case, but you can decide.

Reply:
I feel that extra checks obstruct the intention of the code, so I’d prefer to keep it out.